packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts

```diff
@@ -73,4 +83,106 @@ export class ApplicationExtension<
+        return {
+            routes: serializedRoutes
+        }
+    }
+
+    /**
+     * Returns a map of component names to components that are used to deserialize the extension data.
+     * 
+     * @protected
+     * @returns ComponentMap - The map of component names to components.
+     */
+    protected getComponentMap?(): ComponentMap
+
+    /**
+     * Called on the client to deserialize the extension data that was serialized on the server.
+     * 
+     * @param serializedExtension - The serialized extension data.
+     * @returns DeserializedExtension - The deserialized extension data.
+     * @throws Error if getComponentMap() is not defined.
+     * @throws Error if the deserialized component cannot be found in the component map.
+     */
+    private deserialize(serializedExtension: SerializedExtension): DeserializedExtension {
+        if (!this.getComponentMap) {
+            throw new Error(`${this.getName()}.getRoutes() is async but does not define getComponentMap()`);
+        }
+
+        const componentMap = this.getComponentMap()
+        const routes = serializedExtension.routes.map(
+            ({path, componentName}) => {
+                let component = componentMap[componentName]
+
+                if (!component) {
+                    throw new Error(`${componentName} component could not be deserialized for route with path: ${path}`)
+                }
+
+                return {
+                    path,
+                    exact: true,
+                    component
+                }
+            }
+        )
+        return {routes}
+    }
+
+    private handleAsyncRoutes() {
+        if (!isServerSide) {
+            // Deserialize the routes on the client to ensure the latest routes are loaded on the client
+            this._cachedRoutes = this.deserialize(window.__EXTENSIONS__[this.getName()]).routes;
+        }
+    
+        // Apply caching for the getRoutes method
+        this.applyCacheForMethod('getRoutes', '_cachedRoutes');
+    }
+
+    private isGetRoutesAsync(): boolean {
+        const routes = this.getRoutes();
+        return routes instanceof Promise;
+    }
+
+    private applyCacheForMethod(methodName: string, cacheProperty: string) {
+        const originalMethod = (this as any)[methodName];
+
+        if (typeof originalMethod === 'function') {
+            (this as any)[methodName] = function (this: any, ...args: any[]) {
+                if (this[cacheProperty] !== null) {
+                    return this[cacheProperty];
+                }
+
+                const result = originalMethod.apply(this, args);
+
+                if (result instanceof Promise) {
+                    const promise = result.then((resolved) => (this[cacheProperty] = resolved));
+                    this[cacheProperty] = promise;
+                    return promise;
+                } else {
+                    return (this[cacheProperty] = result);
+                }
+            };
+        }
+    }
```

Comment:
this is the CacheResult decorator logic. I moved it from the decorator to this private method so that it could be applied to getRoutes from the base class.

Now extensions don't need to explicilty add the @CacheResult decorator to the getRoutes function!